### PR TITLE
Support oauth2 authentication

### PR DESCRIPTION
**Pull request by fantapsody**
This PR adds support of authentication methods that generate authentication data dynamically.

- Add a trait `authentication::Authentication` to abstract authentication providers
- Implement `authentication::Authentication` for `connection::Authentication` to make it as a special `authentication::Authentication` implementation.
- Refactor code to use `authentication::Authentication`
  - Initialize the authentication provider at the time of creating the `ConnectionManager`
  - Get the authentication data asynchronously after connected to the remote server
- Implement a token provider
- Implement an (optional) oauth2 provider
  - Basically follows the example of the oauth2 implementation in the pulsar client: https://github.com/apache/pulsar/tree/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/oauth2
  - Cache the access token in memory and request new tokens automatically.
- Add oauth2 support in examples of producer and consumer.

### CODE REVIEW

**Comment by fantapsody** [comment]
Hi @Geal @stearnsc , could you take a look at the PR?

**Review by freeznet — APPROVED** [review]
LGTM

File: src/authentication.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+    use crate::authentication::Authentication;
+    use crate::error::AuthenticationError;
+
+    #[derive(Deserialize, Debug)]
+    struct OAuth2PrivateParams {
+        client_id: String,
+        client_secret: String,
+        client_email: Option<String>,
+        issuer_url: Option<String>,
```
Comment (EronWright) [inline src/authentication.rs:76]: As a stretch goal, please prefer the `issuer_url` within the json file over the configured `issuer_url`.  This makes it possible to support the use of numerous issuers (e.g. for migration purposes).

File: src/authentication.rs [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    }
+
+    #[derive(Deserialize, Debug)]
+    pub struct OAuth2Params {
+        pub issuer_url: String,
+        pub credentials_url: String,
+        pub audience: String,
```
Comment (EronWright) [inline src/authentication.rs]: Please treat `audience` as optional, since it is an Auth0-specific extension.   Azure, in contrast, uses a scope for the same purpose.
